src/core/appcoordinateoperationhandlers.cpp

```diff
@@ -0,0 +1,231 @@
+      const QgsCoordinateReferenceSystem & destinationCrs,
+      const QString & error )
+  {
+    emit coordinateOperationCreationError( sourceCrs, destinationCrs, error );
+  } );
+
+  QgsCoordinateTransform::setCustomMissingGridUsedByContextHandler( [ = ]( const QgsCoordinateReferenceSystem & sourceCrs,
+      const QgsCoordinateReferenceSystem & destinationCrs,
+      const QgsDatumTransform::TransformDetails & desired )
+  {
+    emit missingGridUsedByContextHandler( sourceCrs, destinationCrs, desired );
+  } );
+
+  QgsCoordinateTransform::setFallbackOperationOccurredHandler( [ = ]( const QgsCoordinateReferenceSystem & sourceCrs,
+      const QgsCoordinateReferenceSystem & destinationCrs,
+      const QString & desired )
+  {
+    emit fallbackOperationOccurred( sourceCrs, destinationCrs, desired );
+  } );
+
+  connect( this, &AppMissingGridHandler::missingRequiredGrid, this, &AppMissingGridHandler::onMissingRequiredGrid, Qt::QueuedConnection );
+  connect( this, &AppMissingGridHandler::missingPreferredGrid, this, &AppMissingGridHandler::onMissingPreferredGrid, Qt::QueuedConnection );
+  connect( this, &AppMissingGridHandler::coordinateOperationCreationError, this, &AppMissingGridHandler::onCoordinateOperationCreationError, Qt::QueuedConnection );
+  connect( this, &AppMissingGridHandler::missingGridUsedByContextHandler, this, &AppMissingGridHandler::onMissingGridUsedByContextHandler, Qt::QueuedConnection );
+  connect( this, &AppMissingGridHandler::fallbackOperationOccurred, this, &AppMissingGridHandler::onFallbackOperationOccurred, Qt::QueuedConnection );
+
+  connect( QgsProject::instance(), &QgsProject::cleared, this, [ = ]
+  {
+    mAlreadyWarnedPairsForProject.clear();
+    mAlreadyWarnedBallparkPairsForProject.clear();
+  } );
+}
+
+void AppMissingGridHandler::onMissingRequiredGrid( const QgsCoordinateReferenceSystem &sourceCrs, const QgsCoordinateReferenceSystem &destinationCrs, const QgsDatumTransform::GridDetails &grid )
+{
+  if ( !shouldWarnAboutPair( sourceCrs, destinationCrs ) )
+    return;
+
+  const QString shortMessage = tr( "No transform available between %1 and %2" ).arg( sourceCrs.userFriendlyIdentifier( QgsCoordinateReferenceSystem::ShortString ),
+                               destinationCrs.userFriendlyIdentifier( QgsCoordinateReferenceSystem::ShortString ) );
+
+  QString downloadMessage;
+  if ( !grid.url.isEmpty() )
+  {
+    if ( !grid.packageName.isEmpty() )
+    {
+      downloadMessage = tr( "This grid is part of the “<i>%1</i>” package, available for download from <a href=\"%2\">%2</a>." ).arg( grid.packageName, grid.url );
+    }
+    else
+    {
+      downloadMessage = tr( "This grid is available for download from <a href=\"%1\">%1</a>." ).arg( grid.url );
+    }
+  }
+  QgsMessageLog::logMessage( QStringLiteral( "%1\n%2"  ).arg( shortMessage, downloadMessage ), QStringLiteral( "projection" ) );
```

Comment:
```suggestion
  QgsMessageLog::logMessage( QStringLiteral( "%1\n%2"  ).arg( shortMessage, downloadMessage ), QStringLiteral( tr( "projection" ) ) );
```

?

Reply:
Also, should we add a link to the documentation that explains how to install these files into the correct location?